x-pack/plugins/security/public/session/session_timeout.tsx

```diff
@@ -24,57 +28,168 @@ const GRACE_PERIOD_MS = 5 * 1000;
 
 export interface ISessionTimeout {
-  extend(): void;
+  init(): void;
+  extend(url: string): void;
 }
 
 export class SessionTimeout {
-  private warningTimeoutMilliseconds?: number;
-  private expirationTimeoutMilliseconds?: number;
+  private channel: BroadcastChannel<SessionInfo>;
+  private elector: ReturnType<typeof LeaderElection.create>;
+  private sessionInfo?: SessionInfo;
+  private updateTimer?: number;
+  private warningTimer?: number;
+  private expirationTimer?: number;
   private warningToast?: Toast;
 
   constructor(
-    private sessionTimeoutMilliseconds: number | null,
     private notifications: NotificationsSetup,
     private sessionExpired: ISessionExpired,
     private http: HttpSetup
-  ) {}
+  ) {
+    const name = `${http.basePath.get()}/session_timeout`;
+    this.channel = new BroadcastChannel(name, { webWorkerSupport: false });
+    this.elector = LeaderElection.create(this.channel, {
+      fallbackInterval: 2000,
+      responseTime: 1000,
+    });
+  }
+
+  init() {
+    if (this.isAnon()) {
+      this.elector.die();
+      return;
+    }
+
+    // subscribe to a broadcast channel for session timeout messages
+    // this allows us to synchronize the UX across tabs and avoid repetitive API calls
+    this.channel.onmessage = this.handleSessionInfo;
+    this.elector.awaitLeadership().then(() => {
+      this.updateTimeouts();
+    });
+
+    // Triggers an initial call to the endpoint to get session info;
+    // when that returns, it will set the timeout
+    return this.getSessionInfo();
+  }
 
-  extend() {
-    if (this.sessionTimeoutMilliseconds == null) {
+  extend(url: string) {
+    // ignore these URLs as they are requested by this module
+    if (this.isAnon() || url.endsWith('/session/info') || url.endsWith('/session/extend')) {
       return;
     }
 
-    if (this.warningTimeoutMilliseconds) {
-      window.clearTimeout(this.warningTimeoutMilliseconds);
+    if (this.warningToast) {
+      // the warning is currently showing and the user has clicked elsewhere on the page;
+      // make a new call to get the latest session info
+      return this.getSessionInfo();
+    }
+  }
+
+  private updateTimeouts = () => {
+    const { timeout, isMaximum } = this.getTimeout();
+    if (timeout == null) {
+      return;
+    }
+
+    // cleanup
+    if (this.updateTimer) {
+      window.clearTimeout(this.updateTimer);
+    }
+    if (this.warningTimer) {
+      window.clearTimeout(this.warningTimer);
     }
-    if (this.expirationTimeoutMilliseconds) {
-      window.clearTimeout(this.expirationTimeoutMilliseconds);
+    if (this.expirationTimer) {
+      window.clearTimeout(this.expirationTimer);
     }
     if (this.warningToast) {
       this.notifications.toasts.remove(this.warningToast);
+      this.warningToast = undefined;
     }
-    this.warningTimeoutMilliseconds = window.setTimeout(
+
+    // set timeouts
+    const val = timeout - WARNING_MS - GRACE_PERIOD_MS - 1000;
```

Comment:
The 1000ms is somewhat arbitrary, it just needs to check for updated session info before the warning shows on-screen -- otherwise the user may see the toast flash on and off again.

I'll rework this so it's cleaner.